posthog/hogql/database/schema/channel_type.py

```diff
@@ -0,0 +1,94 @@
+    '$direct',
+    dictGetOrNull(
+        'ga4_channel_definition_dict',
+        'type',
+        cutToFirstSignificantSubdomain(coalesce(properties.$initial_referring_domain, ''))
+    )
+)
+"""
+        ),
+    )
+
+
+def create_initial_channel_type(name: str):
+    return ExpressionField(
+        name=name,
+        expr=parse_expr(
+            """
+multiIf(
+    match(properties.$initial_utm_campaign, 'cross-network'),
+    'Cross Network',
+
+    match(properties.$initial_utm_medium, '^(.*cp.*|ppc|retargeting|paid.*)$'),
+    CASE dictGetOrNull('ga4_channel_definition_dict', 'type', cutToFirstSignificantSubdomain(coalesce(properties.$initial_referring_domain, '')))
+        WHEN 'Shopping' THEN 'Paid Shopping'
+        WHEN 'Search' THEN 'Paid Search'
+        WHEN 'Video' THEN 'Paid Video'
+        WHEN 'Social' THEN 'Paid Social'
+        ELSE multiIf(
+            match(properties.$initial_utm_campaign, '^(.*(([^a-df-z]|^)shop|shopping).*)$'),
+            'Paid Shopping',
+            properties.$initial_utm_medium IN
+                ('display', 'banner', 'expandable', 'interstitial', 'cpm'),
+            'Display',
+            'Paid Other'
+            )
+    END,
+
+    properties.$initial_referring_domain = '$direct' AND (properties.$initial_utm_medium IS NULL OR properties.$initial_utm_medium = ''),
+    'Direct',
+
+    CASE dictGetOrNull('ga4_channel_definition_dict', 'type', cutToFirstSignificantSubdomain(coalesce(properties.$initial_referring_domain, '')))
+        WHEN 'Shopping' THEN 'Organic Shopping'
+        WHEN 'Search' THEN 'Organic Search'
+        WHEN 'Video' THEN 'Organic Video'
+        WHEN 'Social' THEN 'Organic Social'
+        ELSE multiIf(
+            match(properties.$initial_utm_campaign, '^(.*(([^a-df-z]|^)shop|shopping).*)$'),
+            'Organic Shopping',
+            properties.$initial_utm_medium IN
+                ('social', 'social-network', 'social-media', 'sm', 'social network', 'social media'),
+            'Organic Social',
+            match(properties.$initial_utm_campaign, '^(.*video.*)$'),
+            'Organic Video',
+            properties.$initial_utm_medium = 'organic',
+            'Organic Search',
+            properties.$initial_utm_medium IN ('referral', 'app', 'link'),
+            'Referral',
+            properties.$initial_utm_source IN ('email', 'e-mail', 'e_mail', 'e mail')
+                OR properties.$initial_utm_medium IN ('email', 'e-mail', 'e_mail', 'e mail'),
+            'Email',
+            properties.$initial_utm_medium = 'affiliate',
+            'Affiliate',
+            properties.$initial_utm_medium = 'audio',
+            'Audio',
+            properties.$initial_utm_source = 'sms' OR properties.$initial_utm_medium = 'sms',
+            'SMS',
+            match(properties.$initial_utm_medium, '(push$|mobile|notification)')
+                OR properties.$initial_utm_source = 'firebase',
+            'Push',
```

Comment:
Same question for these. This block is looking kind of big. It's small enough to not really cause issues (other than aesthetics under `/debug`?), but can you just add those utm_medium and source combinations straight into the dict, and run the same dictGet on them? Since they're single words, they won't clash with domains... and I see from the JSON the dict also already has things like "Shopping" -> "Shopping". You could also prefix the keys with `domain/`,  `source/` and `medium/` if it makes sense.

Reply:
The reason it's structured the way it is, is to match the order of precedence that GA4 defines. We don't have to do this of course, but for example the way it's defined, if you had
```
utm_medium=social
utm_campaign=video
utm_source=email
```
It's meant to resolve to social because that takes precendence. I think there probably is a way we could do this with the dictionary but we'd lose the precendence.

I don't really know how important this is, it could be that a team of people at Google handcrafted this list as a synthesis of decades of ad-tech knowledge, or it could be a design-by-committee mess that nobody likes much and nobody cares about the edge cases. 🤷  

I might also just be missing something obvious :D

Reply:
I just thought you'd make 3 calls to `dictGet`, one to get the domain match, and two others to get the others. The regexps will likely still need to be inlined... though it might not be as simple as I envisioned, if the order of things matters a lot...

Reply:
This is slightly simplified now, it doesn't stick to the precendence in GA4 as strictly, but it does still work ok with the testing I've done in the wild. Sadly I did need to add more property accesses though, as it needs to check for `gclid` and `gad_source`. Let's merge it and measure perf, there's some obvious things to do like making sure all the properties are materalized (they are not).